osu.Game.Rulesets.Taiko/Difficulty/Utils/IntervalGroupingUtils.cs

```diff
@@ -0,0 +1,63 @@
+// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+using System.Collections.Generic;
+using osu.Framework.Utils;
+
+namespace osu.Game.Rulesets.Taiko.Difficulty.Utils
+{
+    public static class IntervalGroupingUtils
+    {
+        public static List<List<T>> GroupByInterval<T>(IReadOnlyList<T> data, double marginOfError = 5) where T : IHasInterval
+        {
+            var groups = new List<List<T>>();
+            if (data.Count == 0)
+                return groups;
+
+            int i = 0;
+
+            while (i < data.Count)
+            {
+                var group = createGroup(data, ref i, marginOfError);
+                groups.Add(group);
+            }
+
+            return groups;
+        }
+
+        private static List<T> createGroup<T>(IReadOnlyList<T> data, ref int i, double marginOfError) where T : IHasInterval
+        {
+            var children = new List<T> { data[i] };
+            i++;
+
+            for (; i < data.Count - 1; i++)
+            {
+                // An interval change occured, add the current data if the next interval is larger.
```

Comment:
How does this work? Groupings include the next object if the interval *increased* but don't if it *decreased* (as per line 38)? Is there rationale for this decision?

Reply:
The rationale behind it is basically - Even though the spacing is getting a bit wider, it’s still a continuation of the same pattern.

When an interval decreases, it indicates the shift to a faster, or the start of a new rhythmic pattern, whereas the increase implies a slowdown, thus being apart of the same pattern. For example 1/4 - 1/2 is considering easy, whereas 1/4 - 1/6 isn't, we use the grouping logic to indicate the start of a 'new' or in this case a change in difficulty.

Reply:
That sounds fine, but from what I read, it only includes the next object / grouping in this case (ie. it exits the loop) – is this intended or a potential oversight?

Put another way, if the interval decreases twice, this will cause a new grouping to be created, even though there have been no increases in tempo.

Reply:
Seems fine - when i first did the grouping - the double decrease was added as it was considered drastic enough by players for a new grouping to form (i.e. going from 1/6 to 1/3 to 1/1, major slowdowns can cause some discomfort, which was the original intention. Example maps like [Sun and Spirit](https://osu.ppy.sh/beatmapsets/1413087#taiko/2912740) deservedly gain as a result of this.

Reply:
Okay, if that's the intention then it needs to be documented, else it looks like a bug.

Reply:
Another potential bug: the first two objects in a group are *not being compared* for interval checks. Is this intentional..?

Reply:
Another subject to my terrible commenting - I chose this approach to simplify the logic to ensure the behaviour that the first object is the seed, then compare the pairs thereafter - as the second object's interval change would still be considered when looking at the third.